.github/workflows/publish-to-pypi.yml

```
@@ -28,12 +23,51 @@ jobs:
       run: make install && make tests
 
     - name: Poetry bump version, build and publish
+      id: build
       shell: bash
       run: |
         proj_version=$(poetry version -s)
-        if [ $proj_version != $TAG_VERSION ]; then echo "Version $proj_version, defined in pyproject.toml, does not match TAG $TAG_VERSION of this release"; exit 3; fi
+        echo ::set-output name=PROJ_VERSION::$proj_version
         poetry update
         poetry publish --build -u __token__ -p $PYPI_TOKEN
       env:
-        TAG_VERSION: ${{ steps.context.outputs.TAG_VERSION }}
         PYPI_TOKEN: ${{ secrets.PYPI_API_TOKEN }}
+
+    - name: Check if Tag exists 🔍
+      id: check_tag
+      run: |
+        echo "Checking for tag: $PROJ_TAG"
+        if git rev-parse "refs/tags/$PROJ_TAG" >/dev/null 2>&1
+        then
+          echo "Tag $PROJ_TAG already exists."
+          echo "tag_exists=true" >> $GITHUB_OUTPUT
+          exit 0
+        else
+          echo "Tag $PROJ_TAG does not exist."
+          echo "tag_exists=false" >> $GITHUB_OUTPUT
+        fi
+      env:
+        PROJ_TAG: v${{ steps.build.outputs.PROJ_VERSION }}
+
+    - name: Create and Push Tag 🏷️
+      if: steps.check_tag.outputs.tag_exists == 'false'
+      run: |
+        TAG="v${{ steps.extract_version.outputs.version }}"
+        git tag $TAG
+        git config user.name "GitHub Actions"
+        git config user.email "actions@github.com"
+        git remote set-url origin https://x-access-token:${GITHUB_TOKEN}@github.com/${{ github.repository }}
+        git push origin $TAG
```

```suggestion
        git push origin $PROJ_TAG
```

Write permission with the default GH token will also need to be set in the workflow.

---

fixed